### PR TITLE
Add to_list() method to include_directories objects

### PR DESCRIPTION
**Pull request by gabeblack**
This adds a to_list() method on the inc object returned by include_directories. The to_list() method returns a list of strings, where each string is an absolute path (one for the path relative to the source root and the other one for the path relative to build root). If the paths are system paths, then a single path is returned (as-is).

This addition was discussed in this thread: https://github.com/mesonbuild/meson/issues/12012#issue-1811035658

### CODE REVIEW

File: mesonbuild/interpreter/interpreterobjects.py [hunk 1]
```diff
@@ -703,9 +703,18 @@ def subsystem_method(self, args: T.List[TYPE_var], kwargs: TYPE_kwargs) -> str:
             return self.held_object.subsystem
         raise InterpreterException('Subsystem not defined or could not be autodetected.')
 
-
 class IncludeDirsHolder(ObjectHolder[build.IncludeDirs]):
-    pass
+    def __init__(self, target: build.IncludeDirs, interpreter: 'Interpreter'):
```
Comment (bruchar1) [inline mesonbuild/interpreter/interpreterobjects.py:707]:
```suggestion
    def __init__(self, target: build.IncludeDirs, interpreter: Interpreter):
```

File: test cases/common/218 include_dir dot/src/meson.build [hunk 2]
```diff
@@ -1,6 +1,21 @@
...
+source_dir = meson.project_source_root()
+build_dir = meson.project_build_root()
+
+expected = [source_dir / '.', build_dir / '.']
+message('actual: @0@'.format(inc.to_list()))
+message('expected: @0@'.format(expected))
+
+# Use fs.as_posix() because / operator replaces \ with / in the path, but
+# full_path() method is not doing that. This is a pretty inconsistent across all
+# Meson APIs.
```
Comment (bruchar1) [inline test cases/common/218 include_dir dot/src/meson.build:11]: I think this is not the right place to complain about inconsistencies in meson. I would rather something like:

```
Use fs.as_posix() to ensure the results are the same across platforms.
```

File: docs/markdown/snippets/include_directories_to_list.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+## include_directories object now have a to_list() method
+
+The inc object returned by include_directories now has a to_list() method that
+returns a list of strings of absolute paths. Unless the include path is a 
+system path, there will be a path for the source root and one for the build root.
```
Comment (bruchar1) [inline docs/markdown/snippets/include_directories_to_list.md:5]:
```suggestion
The [[@inc]] object returned by [[include_directories]] now has a `to_list()` method that
returns a list of strings of absolute paths. Unless the include path is a 
system path, there will be a path for the source root and one for the build root.
```

File: docs/yaml/objects/inc.yaml [hunk 4]
```diff
@@ -1,3 +1,13 @@
 name: inc
 long_name: Include directories
 description: Opaque wrapper for storing include directories
+
+methods:
+- name: to_list
+  returns: list[str]
+  since: 1.4.0
```
Comment (bruchar1) [inline docs/yaml/objects/inc.yaml:8]:
```suggestion
  since: 1.5.0
```

File: mesonbuild/interpreter/interpreterobjects.py [hunk 5]
```diff
@@ -703,9 +703,18 @@ def subsystem_method(self, args: T.List[TYPE_var], kwargs: TYPE_kwargs) -> str:
...
+        super().__init__(target, interpreter)
+        self.methods.update({'to_list': self.to_list_method})
+
+    @noPosargs
+    @noKwargs
+    @FeatureNew('inc.to_list', '1.4.0')
```
Comment (bruchar1) [inline mesonbuild/interpreter/interpreterobjects.py:713]:
```suggestion
    @FeatureNew('inc.to_list', '1.5.0')
```

File: test cases/common/218 include_dir dot/src/meson.build (hunk 2 above)
Comment (gabeblack) [reply test cases/common/218 include_dir dot/src/meson.build:11]: I wasn't complaining, sorry! I just copy/pasted the same comment from test case #74. I searched to how other test cases addressed the issue and that was the first test case I found. I can change the comment to the language you suggest as well as change the comment in test case 74 to use the same language.

Comment (gabeblack) [reply test cases/common/218 include_dir dot/src/meson.build]: Done. Used the suggested language.

**Comment by bruchar1** [comment]
Could you please squash all your commits into one?

**Comment by gabeblack** [comment]
> Could you please squash all your commits into one?

done.

**Review by bruchar1 — APPROVED** [review]

**Comment by gabeblack** [comment]
@bruchar1 @jpakkane -- what are the next steps to get the approved/merged?

**Comment by gabeblack** [comment]
@bruchar1 @jpakkane -- Are contributions outside of the meson team discouraged? I'm ignorant as to your processes here and am wondering why the radio-silence?

**Comment by gabeblack** [comment]
Was informed that meson can take multiple-months to look at or respond to questions in a PR. Unfortunately, my current work/life balance doesn't allow for me to maintain outside work for long dev cycles, so closing this PR as to not clutter up the PR backlog and make the dev cycle longer for others that are able to do so.

**Comment by WillAyd** [comment]
This is an interesting PR and I think would solve a few usability issues. @gabeblack would you be interested in starting this back up again? If not, perhaps I can take over while retaining authorship credit for you?

**Pull request closed without merging**